shorthand value for CSS gap is not kept as-is in serialized and computed values.

karlcow opened the pull request:
#### ab28ade5ffd5e80451a6a8fc8b0cff43c722ad64
<pre>
shorthand value for CSS gap is not kept as-is in serialized and computed values.
<a href="https://bugs.webkit.org/show_bug.cgi?id=271627">https://bugs.webkit.org/show_bug.cgi?id=271627</a>
<a href="https://rdar.apple.com/125335787">rdar://125335787</a>

Reviewed by Tim Nguyen.

When the shorthand CSS property gap is used, the cssText and computed
declaration can stay in the short form when the values for column and
row are identical.

* LayoutTests/imported/w3c/web-platform-tests/css/css-align/gaps/gap-animation-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-align/gaps/gap-animation-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-align/gaps/gap-animation-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-align/gaps/gap-parsing-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-align/parsing/gap-computed-expected.txt:
* Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
(WebCore::CSSComputedStyleDeclaration::getPropertyValue const):
* Source/WebCore/css/StyleProperties.cpp:
(WebCore::canUseShorthandForLonghand):

Canonical link: <a href="https://commits.webkit.org/276794@main">https://commits.webkit.org/276794@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d47112f46daed78ea2383a518d022ff437f59fa4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45684 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24810 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48262 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48353 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41719 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/47990 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29092 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22206 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37406 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46262 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21896 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39401 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18555 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/19294 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40493 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3728 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/41998 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40825 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50102 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/20673 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17184 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/44510 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21978 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43354 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10151 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22338 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21667 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -250,7 +250,6 @@ static constexpr bool canUseShorthandForLonghand(CSSPropertyID shorthandID, CSSP
     case CSSPropertyColumns:
     case CSSPropertyContainer:
     case CSSPropertyFontSynthesis:
-    case CSSPropertyGap:
```
Is this part covered by WPT? (this is what collapses row-gap + column-gap into gap)

karlcow replied:
oh suddenly you made me wonder about something, aka what is happening with 
`style="row-gap:normal; column-gap:normal"`
insted of just:
`style="gap:normal"`

It's all good. 

```
<div id="target" style="row-gap:normal; column-gap:normal">food</div>

const target = document.getElementById('target');
target.style.cssText;
target.style.getPropertyValue('gap');
window.getComputedStyle(target).gap;


# STP 190
cssText:          "row-gap: normal; column-gap: normal;"
getPropertyValue: "normal"
getComputedStyle: "normal normal"


# This patch (Minibrowser):
cssText:          "gap: normal;"
getPropertyValue: "normal"
getComputedStyle: "normal"


# Firefox 
cssText:          "gap: normal;"
getPropertyValue: "normal"
getComputedStyle: "normal"


# Chrome 
cssText:          "gap: normal;"
getPropertyValue: "normal"
getComputedStyle: "normal"
```

The tests are just covering the value `gap`:
https://github.com/web-platform-tests/wpt/blob/efd5dd8342/css/css-align/parsing/gap-computed.html

```
test_computed_value("gap", "normal");
test_computed_value("gap", "10px");
test_computed_value("gap", "20%");
test_computed_value("gap", "calc(20% + 10px)");
test_computed_value("gap", "calc(-0.5em + 10px)", "0px");
test_computed_value("gap", "calc(0.5em + 10px)", "30px");
```

Maybe our local version of the css-align tests is too old. I should try to import them. 

but this is only testing the `getComputedStyle()` value

The animation tests are also only testing the `getComputedStyle()` values.

Indeed there are no tests for cssText and getPropertyValue.

karlcow replied:
@nt1m I also wonder if I should reimport css-align

karlcow replied:
https://github.com/web-platform-tests/wpt/issues/45334

karlcow replied:
https://github.com/web-platform-tests/wpt/pull/45335

karlcow committed:
shorthand value for CSS gap is not kept as-is in serialized and computed values.
https://bugs.webkit.org/show_bug.cgi?id=271627
rdar://125335787

Reviewed by Tim Nguyen.

When the shorthand CSS property gap is used, the cssText and computed
declaration can stay in the short form when the values for column and
row are identical.

* LayoutTests/imported/w3c/web-platform-tests/css/css-align/gaps/gap-animation-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-align/gaps/gap-animation-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-align/gaps/gap-animation-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-align/gaps/gap-parsing-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-align/parsing/gap-computed-expected.txt:
* Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
(WebCore::CSSComputedStyleDeclaration::getPropertyValue const):
* Source/WebCore/css/StyleProperties.cpp:
(WebCore::canUseShorthandForLonghand):

Canonical link: https://commits.webkit.org/276794@main
